### DISCUSSION TITLE
Can I set the order in which the dataset is entered when I train the dataset?

### DISCUSSION THREAD

**Discussion by mark4653** [ultralytics/ultralytics#18976] [Q&A]
I understand that if you set the batch size to 16 for 64 images, the total images are learned by dividing them into 4 times, 16 each.
At this time, can I decide what image will be in each batch? (For example, when there are images from 1 to 64, I wonder if I can type from 17 to 32 in the first batch and from 1 to 16 in the second batch.)

If I have to modify the source because that feature is not supported, which part should I modify?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @mark4653, thank you for your interest in Ultralytics 🚀! We recommend checking out our [Docs](https://docs.ultralytics.com/) for detailed information on custom training and dataset handling, as many questions like this may already be addressed there.

If this is a ❓ Question related to custom training or dataset order, please provide additional context if possible, including your dataset details, the framework version you are using, and any relevant code snippets to help us better understand your requirements. Additionally, ensure you’re following our [Tips for Best Training Results](https://docs.ultralytics.com/guides/model-training-tips/).

If you aim to modify how batches are ordered, custom changes might be necessary in the source code. For reference, the `DataLoader` implementation typically handles batch formation, so you may need to explore areas like `dataset.py` or the related `DataLoader` configurations. Keep in mind that modifying source code may require some experience with Python and PyTorch.

For any 🐛 Bug Reports related to dataset loading or custom configurations, we kindly ask for a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example/) that demonstrates the behavior.

Join the Ultralytics community to discuss ideas or ask questions directly:
- For real-time help, join our [Discord](https://discord.com/invite/ultralytics) 🎧
- For in-depth discussions, check out our [Discourse](https://community.ultralytics.com/)
- Or connect with others on our [Subreddit](https://reddit.com/r/Ultralytics)

## Upgrade

To ensure the latest features and bug fixes are applied, upgrade your environment to the latest `ultralytics` package with all [requirements](https://github.com/ultralytics/ultralytics/blob/main/pyproject.toml) in a [**Python>=3.8**](https://www.python.org/) environment:

```bash
pip install -U ultralytics
```

## Verified Environments

YOLO can be trained in the following verified environments. These come with all necessary dependencies pre-installed:
- **Notebooks** with free GPU: <a href="https://console.paperspace.com/github/ultralytics/ultralytics"><img src="https://assets.paperspace.io/img/gradient-badge.svg" alt="Run on Gradient"/></a> <a href="https://colab.research.google.com/github/ultralytics/ultralytics/blob/main/examples/tutorial.ipynb"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"></a> <a href="https://www.kaggle.com/models/ultralytics/yolo11"><img src="https://kaggle.com/static/images/open-in-kaggle.svg" alt="Open In Kaggle"></a>
- **Google Cloud** Deep Learning VM. See [GCP Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/google_cloud_quickstart_tutorial/)
- **Amazon** Deep Learning AMI. See [AWS Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/aws_quickstart_tutorial/)
- **Docker Image**. See [Docker Quickstart Guide](https://docs.ultralytics.com/yolov5/environments/docker_image_quickstart_tutorial/) <a href="https://hub.docker.com/r/ultralytics/ultralytics"><img src="https://img.shields.io/docker/pulls/ultralytics/ultralytics?logo=docker" alt="Docker Pulls"></a>

## Status

<a href="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule"><img src="https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml/badge.svg" alt="Ultralytics CI"></a>

If this badge is green, all [Ultralytics CI](https://github.com/ultralytics/ultralytics/actions/workflows/ci.yml?query=event%3Aschedule) tests are passing. CI tests verify correct operation of YOLO [Modes](https://docs.ultralytics.com/modes/) and [Tasks](https://docs.ultralytics.com/tasks/) on macOS, Windows, and Ubuntu every 24 hours and on every commit.

This is an automated response, but don’t worry, an Ultralytics engineer will assist you further soon 🚀!

**Comment by glenn-jocher** [comment]
@mark4653 to customize the batch order during training, you would need to modify the data loading logic. The relevant code is in the `get_dataloader` method where the DataLoader is constructed with `shuffle` parameter control. You can implement a custom sampler or adjust the dataset indices before batching. For specific implementation, we recommend reviewing PyTorch's DataLoader documentation on custom sampling strategies. This functionality isn't natively supported but can be achieved through code modifications.
